apache-Dockerfile-block-1

```diff
@@ -3,7 +3,7 @@ ENV APACHE_ENVVARS $APACHE_CONFDIR/envvars
 
 RUN set -eux; \
 	apt-get update; \
-	apt-get install -y --no-install-recommends apache2; \
+	apt-get install -y --no-install-recommends apache2 runit; \
```

Comment:
I had no idea. AFAIK, Apache doesn't have the functionality to launch / supervise FPM.
I'll remove the supervisor entirely then.

Reply:
I don't think PHP-FPM is capable of running supervised by another process. PHP-FPM is already a process manager and handle thinks like shared memory between processes for cache. Making it run supervised by another process looks impossible to me.

Reply:
If that were the case, it wouldn't be possible to use via `runit`, `systemd`, Docker itself, or even a shell.

However, I can't find anyone crazy enough to try and use it via `FcgidWrapper`, so I'm afraid this might be a dream we can't actually accomplish ATM (which isn't good news for this PR). :disappointed:

Reply:
I agree with @jvasseur. So the main options here are **(1)** have Apache's `mod_fcgid`/`mod_fastcgi` launch FCGI processes, or **(2)** use PHP-FPM to launch FCGI processes.

As stated in the [official docs](https://secure.php.net/manual/en/install.fpm.php), PHP-FPM is an «alternative PHP FastCGI implementation with some additional features (mostly) useful for heavy-loaded sites». According to [some](https://2bits.com/articles/apache-fcgid-acceptable-performance-and-better-resource-utilization.html) [people](https://stackoverflow.com/questions/4526242/what-is-the-difference-between-fastcgi-and-fpm#comment24867165_4526408), PHP-FPM «provides every benefit that fcgid has, with the added advantage of a shared opcode cache for all processes». This would explain why PHP-FPM is generally believed to be more memory-efficient. PHP-FPM is used by default in Fedora.

I tried to find if PHP has a recommendation of PHP-FPM vs standard FCGI, but nothing came up. There is Apache documentation for both approaches, but they are never directly compared. However fcgid would make things easier in the Dockerfile.

Reply:
Well it might be technically possible to use it with `mod_fcgid` if we restrict `mod_fcgid` to run one and only one process (anything else and we would loose all the benefits of php-fpm), but that fells quite hacky.

But I don't think having a supervising process in the image is a good way to go to, if you need to processes you can use two container.

Reply:
> Well it might be technically possible to use it with `mod_fcgid` if we restrict `mod_fcgid` to run one and only one process (anything else and we would loose all the benefits of php-fpm), but that fells quite hacky.

True! I'll do some tests with that, but yeah it's quite hacky.

Simply running `php-fpm` [alongside Apache](https://github.com/docker-library/php/blob/3b46f66f842232171c59f23deea33bb14c5112f5/apache-fpm) without any supervision seems to work well. PHP-FPM is itself a supervisor, so there's a very low risk of it dying (and in that case, it's up to Docker to restart the container).

Reply:
> if you need to processes you can use two container

I usually agree with this, but CGI/FCGI are inter-process interfaces. The FCGI process expects to operate on the same filesystem as the webserver. For some apps it's feasible to run PHP-FPM on a different machine / container (or a farm!) and that's why php:fpm is for. But in other cases, it can be a challenge.

We've tried to migrate some of our images to the dual-container system (php:fpm + apache) and it can be a real nightmare. I think php:apache should provide a transparent option for a more efficient configuration, be it `mod_fcgid`, ZTS or PHP-FPM (even if the latter can be split into 2 containers).